File: bot.js
require('dotenv').config();
const csv = require('csv-parser');
const fs = require('fs');
var handlebars = require('express-handlebars').create({ defaultLayout: 'main' });
var express = require('express');
var app = require('express')();
const http = require('http').createServer(app);
var io = require('socket.io')(http);

// const SQL_FILE = 'db.sql';
const CSV_FILE = 'trivia.csv';

const tmi = require('tmi.js');
// twitch connection config
const options = {
  options: {
    debug: true,
  },
  identity: {
    username: process.env.TWITCH_USER,
    password: process.env.TWITCH_PASS,
  },
  channels: [process.env.CHANNEL],
};
const channel = process.env.CHANNEL;

const client = new tmi.client(options);
// client.connect();
client.on("connected", onConnectedHandler);
client.on("chat", onChatHandler);

var total = 0;
var askedQuestionIds = [];
var askedQuestions = [];
var question = '';
var answer = '';

const mysql = require('mysql');
const { send } = require('process');
// database connection config
var connection = mysql.createConnection({
  host     : process.env.DB_HOST,
  user     : process.env.DB_USER,
  password : process.env.DB_PASS,
  database : 'triviabot',
  port : process.env.DB_PORT,
  multipleStatements: true
});

// connect to database, start front-end, ready bot and database
connection.connect(function(err) {
  if (err) {
    return console.error('*** error: ' + err.message);
  }
  console.log('Connected to the MySQL server.');
  renderWebsite();
  init(); // parce csv, prep database, connect to twitch
});

// prepare the bot and database
async function init() {
  var questionArr = parseCsv(CSV_FILE);
  initializeDatabase();
  await delay(2000);
  loadDatabase(questionArr);
  await delay(2000);
  client.connect();
}

function delay(ms){
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

// ask question once bot has connected to twitch client
function onConnectedHandler(address, port) {
  client.action(channel, "bot has connected");
  askQuestion();
}

// check every message against current answer
function onChatHandler(channel, user, message, self) {
  if (self) {
    return;
  }
  if (message === '!question' || message === '!q') {
    sendQuestion();
  }
  if (message === '!help') {
    client.say(channel, `Enter !question or !q to get the current question. Enter !score to get your score.`);
  }
  if (message === '!score') {
    sendScore(user);
  }
  checkAnswer(user, message);
}

// create triviabot database, questions table, leaderboard table
function initializeDatabase() {
  var sql = "CREATE DATABASE IF NOT EXISTS `triviabot`; \
  USE `triviabot`; \
  SET NAMES utf8; \
  SET character_set_client = utf8mb4; \
  DROP TABLE IF EXISTS `questions`; \
  CREATE TABLE `questions` ( \
    `qid` tinyint(4) NOT NULL AUTO_INCREMENT, \
    `question` varchar(250) NOT NULL, \
    `answer` varchar(100) NOT NULL, \
    PRIMARY KEY (`qid`) \
  ); \
  CREATE TABLE IF NOT EXISTS `leaderboard` ( \
    `user` varchar(120) NOT NULL, \
    `score` tinyint(4) NOT NULL, \
    PRIMARY KEY (`user`) \
  );";
  connection.query(sql, function(err) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
  });
  console.log("Database initialized.");
}

// parse csv for question and answer pairs separated by : and store them in questionArr
function parseCsv(CSV_FILE) {
  var questionArr = [];
  fs.createReadStream(CSV_FILE)
    .pipe(csv({ separator: ':' }))
    .on("data", (row) => {
      console.log(row);
      questionArr.push(row); // appends row to questionArr array
    })
    .on("end", () => {
      console.log("CSV file successfully processed.");
    });
  return questionArr;
}

// load database with questions parsed from csv file
function loadDatabase(questionArr) {
  console.log(questionArr);
  for (let pair of questionArr) {
    var q = pair.question;
    var a = pair.answer;
    var sql = `INSERT INTO questions (question, answer) VALUES ("${q}", "${a}")`;
    connection.query(sql, function(err) {
      if (err) {
        return console.error('*** error: ' + err.message);
      }
    });
  }
  // update total variable with total number of questions loaded into database
  var sql = `SELECT * FROM questions`;
  connection.query(sql, function(err, result) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
    total = result.length;
    console.log("total", total);
  });
  console.log("All questions loaded into database.");
}

// choose and ask new question
async function askQuestion() {
  chooseQuestion();
  await delay(1000);
  sendQuestion();
}

// choose question
async function chooseQuestion() {
  // empty list of askedQuestionIds when all questions have been asked to restart
  if (askedQuestionIds.length === total) {
    askedQuestionIds = [];
  }
  // choose question that has not yet been asked and add to askedQuestionIds
  var added = false;
  while (!(added)) {
    var i = Math.floor(Math.random() * total) + 1;
    if (!(askedQuestionIds.includes(i))) {
      askedQuestionIds.unshift(i);
      added = true;
    }
  }
  // query chosen question from database
  console.log("askedQuestionIds", askedQuestionIds);
  var sql = `SELECT * FROM questions WHERE qid = ${i}`;
  connection.query(sql, function(err, result) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
    // insert question into beginning of list of askedQuestions
    askedQuestions.unshift(result[0]);
    // update current question, answer variables
    question = result[0].question;
    answer = result[0].answer;
  });
  // keep askedQuestions from getting too big
  if (askedQuestions.legnth > 10) {
    askedQuestions.pop();
  }
}

// send current question to twitch chat and front-end
function sendQuestion() {
  if (!(question === '')) {
    client.say(channel, `Question: ${question}`);
    io.emit('current', question);
  }
}

function sendScore(user) {
  var username = user["display-name"] || user["username"];
  var sql = `SELECT score FROM leaderboard WHERE user = "${username}"`;
  connection.query(sql, function(err, result) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
    var score;
    if (result.length == 0) {
      score = 0;
    } else {
      score = result[0].score;
    }
    if (score == 1) {
      client.say(channel, `${username} has ${score} point`);
    } else {
      client.say(channel, `${username} has ${score} points`);
    }
  });
}

// check message with current answer
// if answer is correct, update user's score in leaderboard (addToLeaderboard), update leaderboard in front-end (updateLeaderboard),
// update asked questions in front-end (updatePreviousQestion), choose new question (askQuestion)
function checkAnswer(user, message) {
  if (message.toLowerCase() === answer.toLowerCase()) {
    var username = user["display-name"] || user["username"];
    client.say(channel, `${username} guessed the correct answer: ${answer}`);
    addToLeaderboard(username);
    updateLeaderboard();
    updatePreviousQuestions();
    askQuestion();
  }
}

// increment user's score or adds user to leaderboard database when question is correctly answered
function addToLeaderboard(username) {
  var sql = `INSERT INTO leaderboard (user, score) VALUES ("${username}", 1) ON DUPLICATE KEY UPDATE score = score + 1`;
  connection.query(sql, function(err, result) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
  });
}

// send updated leaderboard to front-end
function updateLeaderboard() {
  connection.query("SELECT * FROM leaderboard ORDER BY score DESC", function(err, result) {
    if (err) {
      return console.error('*** error: ' + err.message);
    }
    io.emit('leaderboard', result);
  });
}

// send updated previous questions to front-end
function updatePreviousQuestions() {
  io.emit('previous', askedQuestions)
}

// connect to front-end
function renderWebsite() {
  app.engine('handlebars', handlebars.engine);
  app.set('view engine', 'handlebars');
  app.use(express.static(__dirname + '/static'));
  app.get('/', function(req, res) {
    res.render('index', {
      layout: 'main',
    });
  });
  http.listen(8000, () => {
    console.log("listening on port 8000");
  });
}

io.on('connection', function(socket) {
  if (askedQuestions.length > 0 && askedQuestions[0].question === question) {
    io.emit('previous', askedQuestions.slice(1));
  } else {
    io.emit('previous', askedQuestions);
  }
  updateLeaderboard();
  io.emit('current', question);
});

File: bot.py
from config import *
import socket
import csv
import random
import sys

# list of (question, answer) tuples
questions = []
# current question #, starting at 1
q_num = 0
# current question
question = ""
# current answer
answer = ""
# flag to close server and bot when game is over
game_over = False

# initialize socket connection
def startSocket():
    server = socket.socket()
    server.connect((HOST, PORT))
    server.send(bytes("PASS " + PASS + "\r\n", "utf-8"))
    server.send(bytes("NICK " + NICK + "\r\n", "utf-8"))
    server.send(bytes("JOIN #" + CHANNEL + "\r\n", "utf-8"))
    return server

# join chat
def joinChat(server):
    while True:
        line = server.recv(2048).decode("utf-8").split("/n").pop()
        if "End of /NAMES list" in line:
            sendMsg(server, "triviabot has joined the chat")
            return

# parse csv file to load questions
def loadQuestions():
    global questions
    with open(FILE, 'r') as csv_file: 
        # create a csv reader object 
        csv_reader = csv.reader(csv_file) 
        # extract each data row one by one 
        for row in csv_reader: 
            questions.append((row[0], row[1]))
        # remove header row
        questions = questions[1:]
        # randomize order of questions
        random.shuffle(questions)
        print("Total number of questions: %d" % len(questions)) 
        # print(questions)

# choose next question from global questions list and update global question, answer, q_num
# vars accordingly. releases question in chat
def chooseQuestion():
    global questions, q_num, question, answer, game_over
    # if all questions have been released, tells chat game is over and updates game_over flag
    if q_num == len(questions):
        sendMsg(server, "Those are all the questions we have! Thanks for playing :^)")
        game_over = True
        return
    curr = questions[q_num]
    question = curr[0]
    answer = curr[1]
    q_num += 1
    # release question in chat
    sendMsg(server, "Question #%d: %s" % (q_num, question))

# pongs server back to avoid timing out
def pong(server, line):
    server.send(bytes(line.replace("PING", "PONG") + "\r\n", "utf-8"))
    print("ponged")

# extracts username
def getUser(line):
    return line.split("!")[0][1:]

# extracts message
def getMsg(line):
    return line.split(":")[2]

# checks if message contains correct answer
def checkAnswer(msg):
    global answer
    if answer.lower() not in msg.lower():
        return False
    return True

# sends message to chat
def sendMsg(server, msg):
    server.send(bytes("PRIVMSG #" + CHANNEL + " :" + msg + "\r\n", "utf-8"))
    print("Sent:", msg)


def run(server):
    loadQuestions()
    server = startSocket()
    joinChat(server)
    chooseQuestion()
    while True:
        # parse through data received from server
        line = server.recv(2048).decode("utf-8").split("\n")
        if len(line) == 2:
            # line is individual chat messages or pings from server
            line = line[0]
            print(line)
            # if server pings, pong back and receive next line
            if "PING" in line:
                pong(server, line)
                continue
            # if user sends a chat message, parse user and message
            else:
                user = getUser(line)
                msg = getMsg(line)
                print(user + ": " + msg)
        # if message contains the answer, tell chat and choose new question
        if checkAnswer(msg):
            sendMsg(server, user + " guessed the correct answer: " + answer)
            chooseQuestion()
        # if all questions have been asnwered, quit bot
        if game_over:
            print("game over")
            server.close()
            sys.exit(0)


if __name__ == "__main__":
    run()
